core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,433 @@
+                              CompletableFuture<Void> completionFuture, SubscriptionOption... options) {
+            requireNonNull(publisher, "publisher");
+            requireNonNull(downstream, "downstream");
+            requireNonNull(executor, "executor");
+            state = head != null ? State.REQUIRE_HEAD : State.REQUIRE_BODY;
+            this.head = head;
+            this.publisher = publisher;
+            this.tail = tail;
+            this.downstream = downstream;
+            this.executor = executor;
+            this.completionFuture = completionFuture;
+            this.options = options;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                close(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (executor.inEventLoop()) {
+                request0(n);
+            } else {
+                executor.execute(() -> request0(n));
+            }
+        }
+
+        private void request0(long n) {
+            if (state == State.DONE) {
+                return;
+            }
+
+            final long oldRequested = requested;
+            if (oldRequested == Long.MAX_VALUE) {
+                return;
+            }
+            if (n == Long.MAX_VALUE) {
+                requested = Long.MAX_VALUE;
+            } else {
+                requested = LongMath.saturatedAdd(oldRequested, n);
+            }
+
+            if (oldRequested > 0) {
+                // SurroundingSubscriber is publishing data.
+                // New requests will be handled by 'publishDownstream(item)'.
+                return;
+            }
+
+            publish();
+        }
+
+        private void publish() {
+            if (state == State.DONE || requested <= 0 && upstreamRequested <= 0) {
+                return;
+            }
+
+            switch (state) {
+                case REQUIRE_HEAD: {
+                    sendHead();
+                    break;
+                }
+                case REQUIRE_BODY: {
+                    if (!subscribed) {
+                        subscribed = true;
+                        publisher.subscribe(this, executor, options);
+                        return;
+                    }
+                    if (upstreamRequested > 0) {
+                        return;
+                    }
+                    final Subscription upstream = this.upstream;
+                    if (upstream != null) {
+                       requestUpstream(upstream);
+                    }
+                    break;
+                }
+                case REQUIRE_TAIL: {
+                    sendTail();
+                    break;
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            assert head != null;
+            publishDownstream(head, true);
```

Comment:
If `head` is published to the downstream, we should assume the ownership of the object is also transferred. If we double release Netty `ByteBuf`, an ISE will be raised.
```suggestion
            final T head = this.head;
            this.head = null;
            publishDownstream(head, true);
```

Reply:
✅ updated